File: scripts/public_sources.py
from __future__ import annotations

import re
from datetime import date, datetime, timedelta
from typing import Dict, List, Optional, Tuple

import pandas as pd
import requests
from bs4 import BeautifulSoup

from bracket_logic import REGION_ORDER, build_bracket_from_region_seed_map, game_pair_key

ESPN_BRACKET_URL = "https://www.espn.com/mens-college-basketball/bracket/_/season/{season}/{season}-ncaa-tournament"
ESPN_SCOREBOARD_URL = "https://site.api.espn.com/apis/site/v2/sports/basketball/mens-college-basketball/scoreboard"
ESPN_TEAMS_URL = "https://site.api.espn.com/apis/site/v2/sports/basketball/mens-college-basketball/teams"


class FetchError(RuntimeError):
    pass


def canonical_name(value: str) -> str:
    text = str(value).strip().lower()
    text = text.replace("&", " and ")
    text = re.sub(r"[^a-z0-9]+", " ", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


def load_alias_map(path: Optional[str]) -> Dict[str, str]:
    if not path:
        return {}
    try:
        df = pd.read_csv(path)
    except FileNotFoundError:
        return {}
    required = {"canonical", "alias"}
    if not required.issubset(df.columns):
        raise ValueError("aliases.csv must have canonical and alias columns")
    mapping: Dict[str, str] = {}
    for _, row in df.iterrows():
        canonical = str(row["canonical"]).strip()
        alias = canonical_name(str(row["alias"]))
        mapping[alias] = canonical
        mapping[canonical_name(canonical)] = canonical
    return mapping


def maybe_alias(name: str, alias_map: Dict[str, str]) -> str:
    norm = canonical_name(name)
    return alias_map.get(norm, str(name).strip())


def parse_matchup_line(text: str) -> Optional[Dict[str, object]]:
    line = " ".join(str(text).replace("\xa0", " ").split())
    if not line:
        return None
    tokens = line.split()
    cutoff = None
    for idx, token in enumerate(tokens):
        if token in {"Mar", "Apr"}:
            cutoff = idx
            break
    if cutoff is not None:
        tokens = tokens[:cutoff]
    seed_positions = [i for i, tok in enumerate(tokens) if tok.isdigit() and 1 <= int(tok) <= 16]
    if len(seed_positions) < 2:
        return None
    p1, p2 = seed_positions[0], seed_positions[1]
    seed_a = int(tokens[p1])
    seed_b = int(tokens[p2])
    team_a = " ".join(tokens[p1 + 1 : p2]).strip()
    team_b = " ".join(tokens[p2 + 1 :]).strip()
    if not team_a or not team_b:
        return None
    return {
        "seed_a": seed_a,
        "seed_b": seed_b,
        "team_a": team_a,
        "team_b": team_b,
        "raw": line,
    }


def fetch_bracket_from_espn(season: int) -> pd.DataFrame:
    url = ESPN_BRACKET_URL.format(season=season)
    response = requests.get(url, timeout=30)
    response.raise_for_status()
    soup = BeautifulSoup(response.text, "html.parser")
    lines = [line.strip() for line in soup.get_text("\n").splitlines() if line.strip()]

    current_section: Optional[str] = None
    current_region: Optional[str] = None
    pending_first_four: Optional[Dict[str, object]] = None
    region_games: Dict[str, List[Dict[str, object]]] = {region: [] for region in REGION_ORDER}
    first_four: List[Tuple[str, str, str, int]] = []

    for line in lines:
        upper = line.upper()
        if line == "First Four":
            current_section = "FIRST_FOUR"
            current_region = None
            continue
        if upper in REGION_ORDER:
            if current_section == "FIRST_FOUR" and pending_first_four:
                first_four.append(
                    (
                        str(pending_first_four["team_a"]),
                        str(pending_first_four["team_b"]),
                        upper,
                        int(pending_first_four["seed"]),
                    )
                )
                pending_first_four = None
            else:
                current_region = upper
                current_section = "MAIN"
            continue
        parsed = parse_matchup_line(line)
        if not parsed:
            continue
        if current_section == "FIRST_FOUR":
            if parsed["seed_a"] != parsed["seed_b"]:
                continue
            pending_first_four = {
                "team_a": parsed["team_a"],
                "team_b": parsed["team_b"],
                "seed": parsed["seed_a"],
            }
            continue
        if current_section == "MAIN" and current_region in REGION_ORDER and len(region_games[current_region]) < 8:
            region_games[current_region].append(parsed)

    if pending_first_four:
        raise FetchError("Bracket scrape found an unassigned First Four game.")
    if any(len(region_games[region]) != 8 for region in REGION_ORDER):
        counts = {region: len(items) for region, items in region_games.items()}
        raise FetchError(f"Bracket scrape failed to capture all regions: {counts}")

    region_seed_map: Dict[str, Dict[int, str]] = {}
    for region in REGION_ORDER:
        seed_map: Dict[int, str] = {}
        for game in region_games[region]:
            seed_map[int(game["seed_a"])] = str(game["team_a"])
            seed_map[int(game["seed_b"])] = str(game["team_b"])
        region_seed_map[region] = seed_map

    return build_bracket_from_region_seed_map(region_seed_map, first_four_pairs=first_four)


def fetch_completed_results(first_day: str, last_day: str, alias_map: Optional[Dict[str, str]] = None) -> Dict[frozenset[str], str]:
    alias_map = alias_map or {}
    start = datetime.strptime(first_day, "%Y-%m-%d").date()
    end = datetime.strptime(last_day, "%Y-%m-%d").date()
    known: Dict[frozenset[str], str] = {}
    day = start
    while day <= end:
        params = {
            "dates": day.strftime("%Y%m%d"),
            "groups": 50,
            "limit": 500,
        }
        response = requests.get(ESPN_SCOREBOARD_URL, params=params, timeout=30)
        response.raise_for_status()
        payload = response.json()
        for event in payload.get("events", []):
            status = event.get("status", {}).get("type", {}).get("name", "")
            if status != "STATUS_FINAL":
                continue
            competitors = event.get("competitions", [{}])[0].get("competitors", [])
            if len(competitors) != 2:
                continue
            names = []
            winner = None
            for comp in competitors:
                team = comp.get("team", {})
                display = team.get("shortDisplayName") or team.get("displayName") or team.get("name")
                if not display:
                    continue
                display = maybe_alias(display, alias_map)
                names.append(display)
                if comp.get("winner"):
                    winner = display
            if len(names) == 2 and winner:
                known[game_pair_key(names[0], names[1])] = winner
        day += timedelta(days=1)
    return known


def fetch_team_logo_map(target_teams: List[str], alias_map: Optional[Dict[str, str]] = None) -> Dict[str, str]:
    alias_map = alias_map or {}
    response = requests.get(ESPN_TEAMS_URL, params={"limit": 1000}, timeout=30)
    response.raise_for_status()
    payload = response.json()

    team_entries = payload.get("sports", [{}])[0].get("leagues", [{}])[0].get("teams", [])
    lookup: Dict[str, str] = {}

    for entry in team_entries:
        team = entry.get("team", {})
        logos = team.get("logos") or []
        if not logos:
            continue
        logo_url = str(logos[0].get("href", "")).strip()
        if not logo_url:
            continue

        candidate_names = {
            team.get("displayName"),
            team.get("shortDisplayName"),
            team.get("nickname"),
            team.get("location"),
            team.get("abbreviation"),
        }
        location = str(team.get("location", "")).strip()
        mascot = str(team.get("name", "")).strip()
        if location and mascot:
            candidate_names.add(f"{location} {mascot}")

        for raw_name in candidate_names:
            if not raw_name:
                continue
            aliased = maybe_alias(str(raw_name), alias_map)
            lookup[canonical_name(aliased)] = logo_url

    result: Dict[str, str] = {}
    for team_name in sorted(set(str(t).strip() for t in target_teams if str(t).strip())):
        key = canonical_name(maybe_alias(team_name, alias_map))
        logo = lookup.get(key)
        if logo:
            result[team_name] = logo

    return result

File: src/server/publicSources.ts
import * as cheerio from "cheerio";

import { buildBracketFromRegionSeedMap, gamePairKey, REGION_ORDER } from "./bracket";
import { BracketGame } from "./types";

const ESPN_BRACKET_URL =
  "https://www.espn.com/mens-college-basketball/bracket/_/season/{season}/{season}-ncaa-tournament";
const ESPN_SCOREBOARD_URL =
  "https://site.api.espn.com/apis/site/v2/sports/basketball/mens-college-basketball/scoreboard";
const ESPN_TEAMS_URL =
  "https://site.api.espn.com/apis/site/v2/sports/basketball/mens-college-basketball/teams";

export class FetchError extends Error {}

export function canonicalName(value: string): string {
  return String(value)
    .trim()
    .toLowerCase()
    .replaceAll("&", " and ")
    .replace(/[^a-z0-9]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

export function loadAliasMap(aliasRows: Array<Record<string, string>>): Record<string, string> {
  const map: Record<string, string> = {};
  for (const row of aliasRows) {
    const canonical = String(row.canonical ?? "").trim();
    const alias = String(row.alias ?? "").trim();
    if (!canonical || !alias) {
      continue;
    }
    map[canonicalName(alias)] = canonical;
    map[canonicalName(canonical)] = canonical;
  }
  return map;
}

export function maybeAlias(name: string, aliasMap: Record<string, string>): string {
  const key = canonicalName(name);
  return aliasMap[key] ?? String(name).trim();
}

interface ParsedMatchupLine {
  seed_a: number;
  seed_b: number;
  team_a: string;
  team_b: string;
}

function parseMatchupLine(text: string): ParsedMatchupLine | null {
  let line = String(text).replaceAll("\u00A0", " ");
  line = line.split(/\s+/).filter(Boolean).join(" ");
  if (!line) {
    return null;
  }

  let tokens = line.split(" ");
  const monthIndex = tokens.findIndex((token) => token === "Mar" || token === "Apr");
  if (monthIndex >= 0) {
    tokens = tokens.slice(0, monthIndex);
  }

  const seedPositions = tokens
    .map((token, index) => ({ token, index }))
    .filter(({ token }) => /^\d+$/.test(token) && Number(token) >= 1 && Number(token) <= 16)
    .map(({ index }) => index);

  if (seedPositions.length < 2) {
    return null;
  }

  const [p1, p2] = seedPositions;
  const seedA = Number(tokens[p1]);
  const seedB = Number(tokens[p2]);
  const teamA = tokens.slice(p1 + 1, p2).join(" ").trim();
  const teamB = tokens.slice(p2 + 1).join(" ").trim();

  if (!teamA || !teamB) {
    return null;
  }

  return {
    seed_a: seedA,
    seed_b: seedB,
    team_a: teamA,
    team_b: teamB,
  };
}

function extractHtmlLines(html: string): string[] {
  const $ = cheerio.load(html);
  const text = $.root().text();
  return text
    .split(/\n+/)
    .map((line) => line.trim())
    .filter(Boolean);
}

export async function fetchBracketFromEspn(season: number): Promise<BracketGame[]> {
  const url = ESPN_BRACKET_URL.replaceAll("{season}", String(season));
  const response = await fetch(url);
  if (!response.ok) {
    throw new FetchError(`Failed to fetch bracket page: HTTP ${response.status}`);
  }

  const html = await response.text();
  const lines = extractHtmlLines(html);

  let currentSection: "FIRST_FOUR" | "MAIN" | null = null;
  let currentRegion: string | null = null;
  let pendingFirstFour: ParsedMatchupLine | null = null;

  const regionGames: Record<string, ParsedMatchupLine[]> = {
    EAST: [],
    WEST: [],
    SOUTH: [],
    MIDWEST: [],
  };

  const firstFour: Array<[string, string, string, number]> = [];

  for (const line of lines) {
    const upper = line.toUpperCase();

    if (line === "First Four") {
      currentSection = "FIRST_FOUR";
      currentRegion = null;
      continue;
    }

    if (REGION_ORDER.includes(upper as (typeof REGION_ORDER)[number])) {
      if (currentSection === "FIRST_FOUR" && pendingFirstFour) {
        firstFour.push([
          pendingFirstFour.team_a,
          pendingFirstFour.team_b,
          upper,
          pendingFirstFour.seed_a,
        ]);
        pendingFirstFour = null;
      } else {
        currentRegion = upper;
        currentSection = "MAIN";
      }
      continue;
    }

    const parsed = parseMatchupLine(line);
    if (!parsed) {
      continue;
    }

    if (currentSection === "FIRST_FOUR") {
      if (parsed.seed_a !== parsed.seed_b) {
        continue;
      }
      pendingFirstFour = parsed;
      continue;
    }

    if (
      currentSection === "MAIN" &&
      currentRegion &&
      REGION_ORDER.includes(currentRegion as (typeof REGION_ORDER)[number]) &&
      regionGames[currentRegion].length < 8
    ) {
      regionGames[currentRegion].push(parsed);
    }
  }

  if (pendingFirstFour) {
    throw new FetchError("Bracket scrape found an unassigned First Four game.");
  }

  const badRegion = Object.entries(regionGames).find(([, games]) => games.length !== 8);
  if (badRegion) {
    const counts = Object.fromEntries(Object.entries(regionGames).map(([region, games]) => [region, games.length]));
    throw new FetchError(`Bracket scrape failed to capture all regions: ${JSON.stringify(counts)}`);
  }

  const regionSeedMap: Record<string, Record<number, string>> = {};
  for (const region of REGION_ORDER) {
    const seedMap: Record<number, string> = {};
    for (const game of regionGames[region]) {
      seedMap[game.seed_a] = game.team_a;
      seedMap[game.seed_b] = game.team_b;
    }
    regionSeedMap[region] = seedMap;
  }

  return buildBracketFromRegionSeedMap(regionSeedMap, firstFour);
}

function toYmdCompact(day: Date): string {
  const yyyy = day.getUTCFullYear();
  const mm = String(day.getUTCMonth() + 1).padStart(2, "0");
  const dd = String(day.getUTCDate()).padStart(2, "0");
  return `${yyyy}${mm}${dd}`;
}

function parseYmd(ymd: string): Date {
  const m = ymd.match(/^(\d{4})-(\d{2})-(\d{2})$/);
  if (!m) {
    throw new Error(`Invalid date format: ${ymd}`);
  }
  const [, y, mo, d] = m;
  return new Date(Date.UTC(Number(y), Number(mo) - 1, Number(d)));
}

export async function fetchCompletedResults(
  firstDay: string,
  lastDay: string,
  aliasMap: Record<string, string> = {},
): Promise<Record<string, string>> {
  const start = parseYmd(firstDay);
  const end = parseYmd(lastDay);
  const known: Record<string, string> = {};

  for (let day = new Date(start); day <= end; day.setUTCDate(day.getUTCDate() + 1)) {
    const params = new URLSearchParams({
      dates: toYmdCompact(day),
      groups: "50",
      limit: "500",
    });

    const response = await fetch(`${ESPN_SCOREBOARD_URL}?${params.toString()}`);
    if (!response.ok) {
      throw new FetchError(`Scoreboard request failed on ${toYmdCompact(day)}: HTTP ${response.status}`);
    }

    const payload = (await response.json()) as {
      events?: Array<{
        status?: { type?: { name?: string } };
        competitions?: Array<{
          competitors?: Array<{
            winner?: boolean;
            team?: {
              shortDisplayName?: string;
              displayName?: string;
              name?: string;
            };
          }>;
        }>;
      }>;
    };

    for (const event of payload.events ?? []) {
      const status = event.status?.type?.name ?? "";
      if (status !== "STATUS_FINAL") {
        continue;
      }

      const competitors = event.competitions?.[0]?.competitors ?? [];
      if (competitors.length !== 2) {
        continue;
      }

      const names: string[] = [];
      let winner = "";

      for (const competitor of competitors) {
        const team = competitor.team ?? {};
        const display = team.shortDisplayName ?? team.displayName ?? team.name;
        if (!display) {
          continue;
        }

        const aliased = maybeAlias(display, aliasMap);
        names.push(aliased);
        if (competitor.winner) {
          winner = aliased;
        }
      }

      if (names.length === 2 && winner) {
        known[gamePairKey(names[0], names[1])] = winner;
      }
    }
  }

  return known;
}

export async function fetchTeamLogoMap(
  targetTeams: string[],
  aliasMap: Record<string, string> = {},
): Promise<Record<string, string>> {
  const response = await fetch(`${ESPN_TEAMS_URL}?limit=1000`);
  if (!response.ok) {
    throw new FetchError(`Team logo request failed: HTTP ${response.status}`);
  }

  const payload = (await response.json()) as {
    sports?: Array<{
      leagues?: Array<{
        teams?: Array<{
          team?: {
            displayName?: string;
            shortDisplayName?: string;
            nickname?: string;
            location?: string;
            abbreviation?: string;
            name?: string;
            logos?: Array<{ href?: string }>;
          };
        }>;
      }>;
    }>;
  };

  const teamEntries = payload.sports?.[0]?.leagues?.[0]?.teams ?? [];
  const lookup: Record<string, string> = {};

  for (const entry of teamEntries) {
    const team = entry.team ?? {};
    const logoUrl = team.logos?.[0]?.href?.trim();
    if (!logoUrl) {
      continue;
    }

    const candidateNames = new Set<string>([
      team.displayName ?? "",
      team.shortDisplayName ?? "",
      team.nickname ?? "",
      team.location ?? "",
      team.abbreviation ?? "",
      [team.location, team.name].filter(Boolean).join(" "),
    ]);

    for (const rawName of candidateNames) {
      if (!rawName) {
        continue;
      }
      const aliased = maybeAlias(rawName, aliasMap);
      lookup[canonicalName(aliased)] = logoUrl;
    }
  }

  const result: Record<string, string> = {};
  for (const team of [...new Set(targetTeams.map((t) => t.trim()).filter(Boolean))]) {
    const key = canonicalName(maybeAlias(team, aliasMap));
    const logo = lookup[key];
    if (logo) {
      result[team] = logo;
    }
  }

  return result;
}
